rag_experiment_accelerator/checkpoint/local_storage_checkpoint.py

```diff
@@ -0,0 +1,52 @@
+import os
+import pickle
+import hashlib
+
+from typing import Any, List, Set
+from rag_experiment_accelerator.checkpoint.checkpoint import Checkpoint
+
+
+class LocalStorageCheckpoint(Checkpoint):
+    """
+    A checkpoint implementation that stores the data in the local file system.
+    """
+
+    def __init__(self, checkpoint_name: str, directory: str = "."):
+        self.checkpoint_location = f"{directory}/checkpoints/{checkpoint_name}"
+        os.makedirs(self.checkpoint_location, exist_ok=True)
+        self.internal_ids: Set[str] = self._get_existing_checkpoint_ids()
+
+    def _has_data(self, id: str, method) -> bool:
+        checkpoint_id = self._build_internal_id(id, method)
+        return checkpoint_id in self.internal_ids
+
+    def _load(self, id: str, method) -> List:
+        file_path = self._get_checkpoint_file_path(id, method)
+        with open(file_path, "rb") as file:
+            data = pickle.load(file)
+            return data
+
+    def _save(self, data: Any, id: str, method):
```

Comment:
How fast is this code? It's going to thrash the file system, isn't it?
How does it work with the multithreading? What happens if I have multiple threads hitting the checkpoint files?
Does this work in AzureML? I assume it does, but wanted to double check

Reply:
There's a tradeoff to doing things this way. This approach was discussed and proposed in the review meeting for this feature. While it does create a large number of files, it simplifies usage, and there is no data loss if the code fails before the checkpoint's save is called. There are no issues with multithreading, and it works relatively fast, especially compared to calling OpenAI again for the same input.
The compute instances in AzureML distribute the data and use ephemeral storage, so there is no point in using it there. Currently, in AzureML, we use the NullCheckpoint implementation, which skips this feature. There is a plan to implement it there using blob storage:  https://github.com/microsoft/rag-experiment-accelerator/issues/186#issuecomment-2050157607

Reply:
Roughly how many files (and storage) is generated per run?

Reply:
It depends on how this is used and on the data, currently it is used for operations on chunks so for every chunk we have 3 calls to the LLM. For the demo data it generated 177 files.